File: app/src/main/java/org/fossasia/openevent/general/ProfileFragment.java
package org.fossasia.openevent.general;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.CardView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import org.fossasia.openevent.general.rest.ApiClient;
import org.fossasia.openevent.general.utils.ConstantStrings;
import org.fossasia.openevent.general.utils.JWTUtils;
import org.fossasia.openevent.general.utils.SharedPreferencesUtil;
import org.json.JSONException;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;


public class ProfileFragment extends Fragment {

    private static final String app="application/vnd.api+json";
    private long userId=-1;
    private TextView firstNameTv;
    private TextView emailTv;
    private ImageView avatarImageView;
    private ProgressBar progressBar;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public ProfileFragment(){
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        String token = SharedPreferencesUtil.getString(ConstantStrings.TOKEN, null);
        if(token == null)
            redirectToLogin();
        token = "JWT "+ token;
        try {
            userId = JWTUtils.getIdentity(token);
            Timber.d("User id is %s", userId);
            ApiClient.setToken(token);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void redirectToLogin() {
        Intent i =new Intent(getActivity(),LoginActivity.class);
        startActivity(i);
    }
    private void redirectToMain() {
        Intent i = new Intent(getActivity(),MainActivity.class);
        startActivity(i);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_profile, container, false);

        firstNameTv = view.findViewById(R.id.first_name_tv);
        emailTv = view.findViewById(R.id.email_tv);
        CardView logout = view.findViewById(R.id.logout_btn);
        logout.setOnClickListener(v -> {
            SharedPreferencesUtil.remove(ConstantStrings.TOKEN);
            ApiClient.setToken(null);
            redirectToMain();
        });
        avatarImageView = view.findViewById(R.id.avatar_image_view);
        progressBar= view.findViewById(R.id.progressHeaderUser);

        progressBar.setIndeterminate(true);

        compositeDisposable.add(ApiClient.getEventApi().getProfile(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(user -> {
                    progressBar.setIndeterminate(false);
                    progressBar.setVisibility(View.GONE);
                    Timber.d("Response Success");
                    firstNameTv.setText(user.getFirstName());
                    emailTv.setText(user.getEmail());

                    Picasso.with(view.getContext())
                            .load(user.getAvatarUrl())
                            .placeholder(R.drawable.ic_person_black_24dp)
                            .transform(new CircleTransform())
                            .into(avatarImageView);
                }, throwable -> {
                    Timber.e(throwable, "Failure");
                }));
        return view;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.dispose();
    }
}

File: app/src/main/java/org/fossasia/openevent/general/ProfileFragment.kt
package org.fossasia.openevent.general

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_profile.view.*
import org.fossasia.openevent.general.rest.ApiClient
import org.fossasia.openevent.general.utils.ConstantStrings
import org.fossasia.openevent.general.utils.JWTUtils
import org.fossasia.openevent.general.utils.SharedPreferencesUtil
import org.json.JSONException
import timber.log.Timber


class ProfileFragment : Fragment() {
    private var userId: Long = -1
    private val compositeDisposable = CompositeDisposable()
    private lateinit var rootView: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var token: String? = SharedPreferencesUtil.getString(ConstantStrings.TOKEN, null)
        if (token == null) {
            redirectToLogin()
            return
        }
        token = "JWT $token"
        try {
            userId = JWTUtils.getIdentity(token).toLong()
            Timber.d("User id is %s", userId)
            ApiClient.setToken(token)
        } catch (e: JSONException) {
            Timber.e(e, "Unable to parse JWT %s", token)
        }
    }

    private fun redirectToLogin() {
        startActivity(Intent(activity, LoginActivity::class.java))
    }

    private fun redirectToMain() {
        startActivity(Intent(activity, MainActivity::class.java))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_profile, container, false)

        rootView.logout.setOnClickListener { _ ->
            SharedPreferencesUtil.remove(ConstantStrings.TOKEN)
            ApiClient.setToken(null)
            redirectToMain()
        }

        rootView.progressBar.isIndeterminate = true

        compositeDisposable.add(ApiClient.eventApi.getProfile(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ user ->
                    rootView.progressBar.isIndeterminate = false
                    rootView.progressBar.visibility = View.GONE
                    Timber.d("Response Success")
                    rootView.name.text = "${user.firstName} ${user.lastName}"
                    rootView.email.text = user.email

                    Picasso.get()
                            .load(user.avatarUrl)
                            .placeholder(R.drawable.ic_person_black_24dp)
                            .transform(CircleTransform())
                            .into(rootView.avatar)
                }) { throwable -> Timber.e(throwable, "Failure") })
        return rootView
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.dispose()
    }

}